File: 003-LCS/python/lcs_bottom_up.py
# Strings to find LCS of
str1, str2 = 'apple', 'apollo'

# Find length of LCS of str1 and str2 bottom up
def getLengthLCS(n1, n2):
    # Create a 2d array of size (n1 + 1) X (n2 + 1)
    lengthOfLCS = [[None for j in range(n2 + 1)] for i in range(n1 + 1)]

    # Initialize the base cases
    for i in range(n1 + 1): lengthOfLCS[i][0] = 0
    for j in range(n2 + 1): lengthOfLCS[0][j] = 0

    # Iterate till result is found
    for i in range(1, n1 + 1):
        for j in range(1, n2 + 1):
            if str1[i - 1] == str2[j - 1]:
                lengthOfLCS[i][j] = lengthOfLCS[i - 1][j - 1] + 1
            else:
                lengthOfLCS[i][j] = max(lengthOfLCS[i - 1][j], lengthOfLCS[i][j - 1])

    return lengthOfLCS[n1][n2]

if __name__ == '__main__':
    print(getLengthLCS(len(str1), len(str2)))

File: 003-LCS/cpp/lcs_bottom_up.cpp
#include<bits/stdc++.h>
using namespace std;

// Strings to find LCS of
string str1 = "apple", str2 = "apollo";

// Method to find length of LCS of two strings
int getLengthLCS(int n1, int n2) {
    // Create array of dimention (n1 + 1) X (n2 + 1)
    int **lengthOfLCS = new int*[n1 + 1];
    for (int i = 0; i <= n1; i++) {
        lengthOfLCS[i] = new int[n2 + 1];
    }

    // Initialize the base case
    for (int i = 0; i <= n1; i++) lengthOfLCS[i][0] = 0;
    for (int j = 0; j <= n2; j++) lengthOfLCS[0][j] = 0;

    // Keep iterating until result is found
    for (int i = 1; i <= n1; i++) {
        for (int j = 1; j <= n2; j++) {
            if (str1[i - 1] == str2[j - 1]) {
                lengthOfLCS[i][j] = lengthOfLCS[i - 1][j - 1] + 1;
            }
            else {
                lengthOfLCS[i][j] = max(lengthOfLCS[i - 1][j], lengthOfLCS[i][j - 1]);
            }
        }
    }

    return lengthOfLCS[n1][n2];
}

int main() {
    cout << getLengthLCS(str1.length(), str2.length()) << endl;

    return 0;
}
